### DISCUSSION TITLE
How to access metadata while invoking service?

### DISCUSSION THREAD

**Discussion by ramesh-techgrit** [statelyai/xstate#2770] [Q&A]
I am designing a messaging workflow state machine , where i have different messages(sms and mms) configured for different days, and i am sending the messages after checking the guard/conditions. Currently i have written the message send as a action, but as this is a asynchronous process i was trying to change it to a callback with invoke service. In action i am getting the message from meta, but in service i couldnt find a way to access metadata. Also i am persisting the state to the db in onTransition which also not working as expected when changed to service. Please help , visualizer link given below.

https://stately.ai/viz/01407072-0e59-45c6-bc6d-ca31d0302e5b

**Comment by rameshrajappan** [comment]
Can anyone please help?

**Reply by davidkpiano** [reply]
Where do you need to access metadata in the service? In `canSendMessage`?

**Reply by ramesh-techgrit** [reply]
No i am currently sending message using an action "sendSMSMessage" . I want to convert it to a invoke/service so that i can handle callback/Promise and check if there are any errors and update the machine state to database accordingly.

**Reply by ramesh-techgrit** [reply]
But when i change the action "sendSMSMessage" to an invoke method ,i am not able to access the meta argument.

**Reply by Andarist** [reply]
Currently, you don't get access to `state` as part of the "invoke meta" (3rd param of the function).

However, I think that `state` that is available there for actions and guards is often misunderstood and I'm trying to convince David that we should remove it 😅 Don't worry though! It's not like I want to completely remove the access to the information that can be often found there. I would just love to provide it in a more reliable way and I believe that my idea would satisfy your needs. When looking at your code I see that you don't actually need access to the state but rather to the information put on the **state node** (in this situation you actually want to access the meta put on the parent state node). This shows in your implementation as you need to unpack the desired meta from the `state` object in a quirky way:
```js
const metaKey = Object.keys(state.meta)[0];
const metaData = state.meta[metaKey];
```

It's not wrong per se (there is no easier way to do this right now) - but it shows how convoluted using this can actually be and that changing how we provide this sort of information could simplify this greatly. The drawback of the current solution is that this has the potential of breaking whenever you add a parallel state to your machine and that's not good. An unrelated change should never have such a severe impact on the implementation that is somewhere else. Almost nobody would predict that this code might require adjusting when adding a parallel state.

I might also have some good news for you. You can get access to this information if you "duplicate" it on the `invoke.meta`, like here;
https://codesandbox.io/s/aged-butterfly-8oup7?file=/src/index.js:65-320

Note that you won't see this working in the visualizer right now because visualizer is running on a slightly older version of XState and `invoke.meta` has been added in [4.26.0](https://github.com/statelyai/xstate/releases/tag/xstate%404.26.0). I will update the XState version in the visualizer in the moment so it should start working there soon too

**Reply by ramesh-techgrit** [reply]
Thank you so much.
